pkg/prom/instance/instance.go

```diff
@@ -390,6 +393,81 @@ func (i *Instance) initialize(ctx context.Context, reg prometheus.Registerer) er
 	return nil
 }
 
+// Update accepts a new Config for the Instance and will dynamically update any
+// running Prometheus components with the new values from Config. Update will
+// return an ErrInvalidUpdate if the Update could not be applied.
+func (i *Instance) Update(c Config) error {
+	// NOTE: you must get a lock on componentMtx first, since initialize will
+	// obtain a lock on cfgMutex and will deadlock if that is already locked
+	// by Update.
+	i.componentMtx.Lock()
```

Comment:
These locks are here for pretty much the same purpose, I think we could unify them.